internal/expr/parser.go

```diff
@@ -734,21 +733,21 @@ func (p *Parser) parseTargetTypes() (types []memberAccessor, parentheses bool, o
 // parseOutputExpr requires that the ampersand before the identifiers must
 // be followed by a name byte.
 func (p *Parser) parseOutputExpr() (*outputExpr, bool, error) {
-	start := p.pos
+	cp := p.save()
+	p.currentExprStart = p.pos
```

Comment:
Having `currentExprStart` set here and used in `add` feels a bit weird to me. One of the abstractions I don't really want to break with the parser is that we do all the setting of the `p.` variables in the functions directly related to the operation of the parser i.e. `Parse`, `advance` and `add` and not elsewhere. 

I think even just abstracting this to a separate function rather than setting it directly here would be good. Then that function can sit beside `add` in the code to make things a bit more readable. This also has the benefit of forcing some documentation.

Reply:
I agree `add` could be refactored but I think setting the `currentExprStart` in `advance` is very difficult to reason about. It makes much more sense to set it when you are finding an expression and only if you have found one. Else you get jumps for now reason every time you call advance which makes other ways of parsing (e.g. recursive) very hacky.

Conceptually I will like to set the expression start only when there is an expression (what we are achieving with the checkpoint there). What we can do is change the `add` function and separate it into one that adds the bypass and is documented as such. At the same time I dont want to add the part inside the `parseOutputExpr` because you might want to parse and not add it (e.g. to collect a list and discard it if necessary).